src/gui/shader_manager.h

```diff
@@ -0,0 +1,180 @@
+	// "how people experienced the game at the time of release", and
+	// prioritising the most probable developer intent.)
+	//
+	AutoGraphicsStandard,
+
+	// Machine-based adaptive CRT shader mode.
+	// Enabled via the 'crt-machine-auto' magic 'glshader' setting.
+	//
+	// This mode emulates a computer (machine) equipped with the configured
+	// video adapter and a matching monitor. The auto-switching picks the most
+	// approriate shader variant for the adapter & monitor combo (Hercules,
+	// CGA, EGA, (S)VGA, etc.) for a given viewport resolution.
+	//
+	// E.g., CGA and EGA modes on an emulated VGA adapter type will always use
+	// 'crt/vga-*' shaders, on an EGA adapter always the 'crt/ega-*' shaders,
+	// and so on.
+	//
+	AutoMachine,
+
+	// 15 kHz arcade / home computer monitor adaptive CRT shader mode.
+	// Enabled via the 'crt-machine-arcade' magic 'glshader' setting.
+	//
+	// This basically forces single-scanning of all double-scanned VGA modes
+	// and no pixel-doubling in all modes to achieve a somewhat less sharp
+	// look with more blending and "rounder" pixels than what you'd get on a
+	// typical sharp EGA/VGA PC monitor.
+	//
+	// This is by no means "authentic", but a lot of fun with certain games,
+	// plus it allows you to play DOS ports of Amiga games or other 16-bit
+	// home computers with a single-scanned 15 kHz monitor look.
+	//
+	AutoArcade
+};
+
+struct ShaderSettings {
+	bool use_srgb_texture        = false;
+	bool use_srgb_framebuffer    = false;
+	bool force_single_scan       = false;
+	bool force_no_pixel_doubling = false;
+};
+
+struct ShaderInfo {
+	std::string name        = {};
+	ShaderSettings settings = {};
+	bool is_adaptive        = false;
+};
+
+// Shader manager that picks the best shader to use based on various criteria.
+// Its main function is to handle shader auto-switching in the adaptive CRT
+// shader modes.
+//
+// Usage:
+//
+// - Notify the shader manager about changes that could potentially trigger
+//   shader switching with the `Notify*` methods.
+//
+// - Query information about the current shader with `GetCurrentShaderInfo()`.
+//
+// - Fetch the source of the current shader with `GetCurrentShaderSource()`.
+//
+// The caller is responsible for compiling the shader source to the OpenGL
+// backend and to implement lazy shader switching (only when the shader has
+// really changed).
+//
+class ShaderManager {
+public:
+	ShaderManager()  = default;
+	~ShaderManager() = default;
+
+	std::deque<std::string> InventoryShaders() const;
+
+	std::string MapShaderName(const std::string& name) const;
+
+	void NotifyGlshaderSettingChanged(const std::string& shader_name);
+
+	void NotifyRenderParametersChanged(const uint16_t canvas_width,
+	                                   const uint16_t canvas_height,
+	                                   const VideoMode& video_mode);
+
+	ShaderInfo GetCurrentShaderInfo() const;
+	std::string GetCurrentShaderSource() const;
```

Comment:
Suggest returning `const std::string&`, the caller can store a value  if they want to work with it:

```c++
std::string shader_copy = GetCurrentShaderSource();
```

Reply:
Sure, that's better.